internal/tui/components/chat/messages/tool.go

```diff
@@ -832,34 +802,134 @@ func (m *toolCallCmp) SetSize(width int, height int) tea.Cmd {
-// Returns true if the tool call is not finished or if the result doesn't match the call ID.
+// Returns true if the tool call is not finished and not in a terminal state.
 func (m *toolCallCmp) shouldSpin() bool {
-	return !m.call.Finished && !m.cancelled
+	// Tools should only spin when in non-final states
+	return m.call.State.IsNonFinalState()
+}
+
+// UpdateAnimationState updates the animation state based on current tool call and result data.
+// This is the single source of truth for determining animation behavior.
+func (m *toolCallCmp) UpdateAnimationState() {
+	// Get effective display state considering both tool call state and results
+	effectiveState := m.getEffectiveDisplayState()
+	m.animationState = effectiveState.ToAnimationState()
 }
 
 // Spinning returns whether the tool call is currently showing a loading animation
-func (m *toolCallCmp) Spinning() bool {
-	if m.spinning {
+func (m *toolCallCmp) IsAnimating() bool {
+	if m.animationState.IsActive() {
 		return true
 	}
 	for _, nested := range m.nestedToolCalls {
 		if nested.Spinning() {
 			return true
 		}
 	}
-	return m.spinning
+	return false
+}
+
+// Spinning returns whether tool call is currently showing a loading animation
+// Deprecated: Use IsAnimating() instead for better type safety
+func (m *toolCallCmp) Spinning() bool {
+	return m.animationState.IsActive()
 }
 
 func (m *toolCallCmp) ID() string {
 	return m.call.ID
 }
 
-// SetPermissionRequested marks that a permission request was made for this tool call
-func (m *toolCallCmp) SetPermissionRequested() {
-	m.permissionRequested = true
+// SetToolCallState sets the tool call state
+func (m *toolCallCmp) SetToolCallState(state enum.ToolCallState) {
+	m.call.State = state
+	m.updateAnimationForState()
 }
 
-// SetPermissionGranted marks that permission was granted for this tool call
-func (m *toolCallCmp) SetPermissionGranted() {
-	m.permissionGranted = true
+// updateAnimationForState reconfigures animation based on current tool call state
+func (m *toolCallCmp) updateAnimationForState() {
+	t := styles.CurrentTheme()
+
+	// Configure animation based on tool call state
+	switch m.call.State {
+	case enum.ToolCallStatePending:
+		// State 1 (Unstarted): static - no animation
+		m.anim = anim.New(anim.Settings{
+			Size:        15,
+			Label:       "Pending",
+			GradColorA:  t.FgMuted,
+			GradColorB:  t.FgMuted,
+			LabelColor:  t.FgSubtle,
+			CycleColors: false,
+		})
+	case enum.ToolCallStatePermissionPending:
+		// State 2 (Awaiting permission): timer counts up every 1s
+		m.anim = anim.New(anim.Settings{
+			Size:        15,
+			Label:       "Requesting Permission",
+			GradColorA:  t.Paprika,
+			GradColorB:  t.Paprika,
+			LabelColor:  t.FgBase,
+			CycleColors: false,
+		})
+	case enum.ToolCallStatePermissionApproved:
+		// State 2.5 (Permission approved): transitioning to running
+		m.anim = anim.New(anim.Settings{
+			Size:        15,
+			Label:       "Running",
+			GradColorA:  t.GreenDark,
```

Comment:
Consider: Moving this funtion to `internal/enum/AnimationState.go` and add `ToolCallState` as a function parameter.

Contra: `m.UpdateAnimationState()` call at the end.